motr/idx_dix.c

```diff
@@ -554,12 +554,33 @@ static void dix_build(const struct m0_op_idx *oi,
 	out->dd_fid = *OI_IFID(oi);
 	/* Pool version and layout type which are passed by consumers like S3 */
 	if (M0_IN(opcode, (M0_IC_GET, M0_IC_PUT, M0_IC_DEL, M0_IC_NEXT))) {
-		if (idx->in_attr.idx_layout_type == DIX_LTYPE_DESCR) {
+		if ((idx->in_attr.idx_layout_type == DIX_LTYPE_DESCR) 
+		    && (m0_fid_is_set(&idx->in_attr.idx_pver))
+		    && (m0_fid_is_valid(&idx->in_attr.idx_pver))) {
 			M0_LOG(M0_DEBUG, "Opcode: %u, DIX pool version:"FID_F"",
 			       opcode, FID_P(&idx->in_attr.idx_pver));
+
 			out->dd_layout.dl_type = DIX_LTYPE_DESCR;
-			out->dd_layout.u.dl_desc.ld_pver = idx->in_attr.idx_pver;
+			m0_dix_ldesc_init(&out->dd_layout.u.dl_desc,
+					  &(struct m0_ext) { .e_start = 0,
+					  .e_end = IMASK_INF }, 1,
+					  HASH_FNC_CITY,
+					  &idx->in_attr.idx_pver);
 		}
+	} else if (M0_IN(opcode, (M0_EO_CREATE))) {
+		/*
+		 * Use default layout for all indices:
+		 * - city hash function;
+		 * - infinity identity mask (use key as is);
+		 * - default pool version (the same as for root index).
+		 * In future client user will be able to pass layout as an 
+		 * argument.
+		 */
+		out->dd_layout.dl_type = DIX_LTYPE_DESCR;
+		m0_dix_ldesc_init(&out->dd_layout.u.dl_desc,
+				  &(struct m0_ext) { .e_start = 0,
+				  .e_end = IMASK_INF }, 1, HASH_FNC_CITY,
+				  &idx->in_attr.idx_pver);
```

Comment:
The logic seems fine to me.
But the code of these two conditions seems a bit duplicated.

Reply:
Correct. But if conditions are different and dix_build() is common function for other operations.

Reply:
OK